File: frontend/src/lib/api/as-path-list.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface AsPathListRule {
  rule_number: number;
  description?: string | null;
  action: string;  // permit|deny
  regex?: string | null;
}

export interface AsPathList {
  name: string;
  description?: string | null;
  rules: AsPathListRule[];
}

export interface AsPathListConfig {
  as_path_lists: AsPathList[];
  total: number;
}

export interface AsPathListCapabilities {
  version: string;
  features: {
    basic: { supported: boolean; description: string };
    rules: { supported: boolean; description: string };
    actions: { supported: boolean; description: string };
  };
  version_notes: {
    identical_versions: string;
  };
  device_name?: string;
}

export interface AsPathListBatchOperation {
  op: string;
  value?: string;
}

export interface AsPathListBatchRequest {
  name: string;
  rule_number?: number;
  operations: AsPathListBatchOperation[];
}

// ============================================================================
// API Service
// ============================================================================

class AsPathListService {
  /**
   * Get capabilities based on VyOS version
   */
  async getCapabilities(): Promise<AsPathListCapabilities> {
    return apiClient.get<AsPathListCapabilities>("/vyos/as-path-list/capabilities");
  }

  /**
   * Get all AS path lists configuration
   */
  async getConfig(refresh: boolean = false): Promise<AsPathListConfig> {
    return apiClient.get<AsPathListConfig>("/vyos/as-path-list/config", {
      refresh: refresh.toString(),
    });
  }

  /**
   * Refresh the cached configuration
   */
  async refreshConfig(): Promise<unknown> {
    return apiClient.post("/vyos/config/refresh");
  }

  /**
   * Execute batch operations
   */
  async batchConfigure(request: AsPathListBatchRequest): Promise<unknown> {
    const result = await apiClient.post("/vyos/as-path-list/batch", request);
    await this.refreshConfig();
    return result;
  }

  /**
   * Delete an entire AS path list
   */
  async deleteAsPathList(name: string): Promise<unknown> {
    const operations: AsPathListBatchOperation[] = [];
    operations.push({ op: "delete_as_path_list" });

    return this.batchConfigure({
      name,
      rule_number: 0, // Not used for delete_as_path_list
      operations,
    });
  }

  /**
   * Delete a specific rule from an AS path list and renumber remaining rules to close gaps
   */
  async deleteRule(name: string, ruleNumber: number): Promise<unknown> {
    // Get current configuration
    const config = await this.getConfig(true);
    const asPathList = config.as_path_lists.find(apl => apl.name === name);

    if (!asPathList) {
      throw new Error(`AS path list ${name} not found`);
    }

    // Get remaining rules (excluding the one being deleted)
    const remainingRules = asPathList.rules.filter(r => r.rule_number !== ruleNumber);

    if (remainingRules.length === 0) {
      // If no rules left, just delete the rule
      const operations: AsPathListBatchOperation[] = [];
      operations.push({ op: "delete_rule" });

      return this.batchConfigure({
        name,
        rule_number: ruleNumber,
        operations,
      });
    }

    // Sort remaining rules by their current number
    const sortedRules = remainingRules.sort((a, b) => a.rule_number - b.rule_number);

    // Renumber sequentially starting from the lowest existing number
    // This closes gaps: if you have 105, 106, 107, 108 and delete 106,
    // result will be 105, 106 (was 107), 107 (was 108)
    const startingNumber = sortedRules[0].rule_number;
    const reorderRules = sortedRules.map((rule, index) => ({
      old_number: rule.rule_number,
      new_number: startingNumber + index,
      rule_data: rule,
    }));

    // Check if any rule actually needs renumbering
    const needsReorder = reorderRules.some(r => r.old_number !== r.new_number);

    if (!needsReorder) {
      // No gaps to close, just delete the rule directly
      const operations: AsPathListBatchOperation[] = [];
      operations.push({ op: "delete_rule" });

      return this.batchConfigure({
        name,
        rule_number: ruleNumber,
        operations,
      });
    }

    // Use reorder endpoint which will delete all rules (including the one we want to delete)
    // and recreate them with new sequential numbers
    return this.reorderRules(name, reorderRules);
  }

  /**
   * Helper: Create a new AS path list with a rule
   */
  async createAsPathList(name: string, description: string | null, rule: Partial<AsPathListRule>): Promise<unknown> {
    const operations: AsPathListBatchOperation[] = [];

    // Create AS path list
    operations.push({ op: "set_as_path_list" });

    // Add description
    if (description) {
      operations.push({ op: "set_as_path_list_description", value: description });
    }

    // Create rule
    operations.push({ op: "set_rule" });

    // Rule description
    if (rule.description) {
      operations.push({ op: "set_rule_description", value: rule.description });
    }

    // Rule action
    if (rule.action) {
      operations.push({ op: "set_rule_action", value: rule.action });
    }

    // Rule regex
    if (rule.regex) {
      operations.push({ op: "set_rule_regex", value: rule.regex });
    }

    return this.batchConfigure({
      name,
      rule_number: rule.rule_number,
      operations,
    });
  }

  /**
   * Helper: Update an existing AS path list or rule
   */
  async updateAsPathList(
    name: string,
    originalAsPathList: AsPathList,
    description: string | null,
    rule?: Partial<AsPathListRule>,
    ruleNumber?: number
  ): Promise<unknown> {
    const operations: AsPathListBatchOperation[] = [];

    // Update description
    if (description !== originalAsPathList.description) {
      if (description) {
        operations.push({ op: "set_as_path_list_description", value: description });
      } else if (originalAsPathList.description) {
        operations.push({ op: "delete_as_path_list_description" });
      }
    }

    // If updating a rule
    if (rule && ruleNumber !== undefined) {
      const originalRule = originalAsPathList.rules.find(r => r.rule_number === ruleNumber);

      // Rule description
      if (rule.description !== originalRule?.description) {
        if (rule.description) {
          operations.push({ op: "set_rule_description", value: rule.description });
        } else if (originalRule?.description) {
          operations.push({ op: "delete_rule_description" });
        }
      }

      // Rule action
      if (rule.action && rule.action !== originalRule?.action) {
        operations.push({ op: "set_rule_action", value: rule.action });
      }

      // Rule regex
      if (rule.regex !== originalRule?.regex) {
        if (rule.regex) {
          operations.push({ op: "set_rule_regex", value: rule.regex });
        } else if (originalRule?.regex) {
          operations.push({ op: "delete_rule_regex" });
        }
      }
    }

    return this.batchConfigure({
      name,
      rule_number: ruleNumber,
      operations,
    });
  }

  /**
   * Helper: Add a new rule to existing AS path list
   */
  async addRule(name: string, rule: Partial<AsPathListRule>): Promise<unknown> {
    const operations: AsPathListBatchOperation[] = [];

    // Create rule
    operations.push({ op: "set_rule" });

    // Rule description
    if (rule.description) {
      operations.push({ op: "set_rule_description", value: rule.description });
    }

    // Rule action
    if (rule.action) {
      operations.push({ op: "set_rule_action", value: rule.action });
    }

    // Rule regex
    if (rule.regex) {
      operations.push({ op: "set_rule_regex", value: rule.regex });
    }

    return this.batchConfigure({
      name,
      rule_number: rule.rule_number,
      operations,
    });
  }

  /**
   * Update an existing rule
   */
  async updateRule(name: string, ruleNumber: number, rule: Partial<AsPathListRule>): Promise<unknown> {
    const operations: AsPathListBatchOperation[] = [];

    // Rule description
    if (rule.description !== undefined) {
      if (rule.description) {
        operations.push({ op: "set_rule_description", value: rule.description });
      } else {
        operations.push({ op: "delete_rule_description" });
      }
    }

    // Rule action
    if (rule.action) {
      operations.push({ op: "set_rule_action", value: rule.action });
    }

    // Rule regex
    if (rule.regex !== undefined) {
      if (rule.regex) {
        operations.push({ op: "set_rule_regex", value: rule.regex });
      } else {
        operations.push({ op: "delete_rule_regex" });
      }
    }

    return this.batchConfigure({
      name,
      rule_number: ruleNumber,
      operations,
    });
  }

  /**
   * Reorder AS path list rules
   */
  async reorderRules(asPathListName: string, rules: Array<{ old_number: number; new_number: number; rule_data: AsPathListRule }>): Promise<unknown> {
    const result = await apiClient.post("/vyos/as-path-list/reorder", {
      as_path_list_name: asPathListName,
      rules: rules,
    });
    await this.refreshConfig();
    return result;
  }
}

export const asPathListService = new AsPathListService();

File: backend/routers/as_path_list/as_path_list.py
"""
AS Path List Router

API endpoints for managing VyOS AS path list configuration.
Supports version-aware configuration for VyOS 1.4 and 1.5 (identical feature sets).
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders import AsPathListBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
from utils.router_helpers import load_vyos_capabilities
import inspect

router = APIRouter(prefix="/vyos/as-path-list", tags=["as-path-list"])

# Stub functions for backwards compatibility with app.py
def set_device_registry(registry):
    """Legacy function - no longer used."""
    pass


def set_configured_device_name(name):
    """Legacy function - no longer used."""
    pass


# ============================================================================
# Pydantic Models
# ============================================================================


class AsPathListRule(BaseModel):
    """AS path list rule"""
    rule_number: int
    description: Optional[str] = None
    action: str = "permit"  # permit|deny
    regex: Optional[str] = None


class AsPathList(BaseModel):
    """Complete AS path list configuration"""
    name: str
    description: Optional[str] = None
    rules: List[AsPathListRule] = []


class AsPathListConfig(BaseModel):
    """Response containing all AS path lists"""
    as_path_lists: List[AsPathList] = []
    total: int = 0


class AsPathListBatchOperation(BaseModel):
    """Single operation in a batch request"""
    op: str = Field(..., description="Operation name")
    value: Optional[str] = Field(None, description="Operation value")


class AsPathListBatchRequest(BaseModel):
    """Model for batch configuration"""
    name: str = Field(..., description="AS path list name")
    rule_number: Optional[int] = Field(None, description="Rule number (optional)")
    operations: List[AsPathListBatchOperation]


class ReorderRuleItem(BaseModel):
    """Single rule in a reorder request"""
    old_number: int = Field(..., description="Original rule number")
    new_number: int = Field(..., description="New rule number after reorder")
    rule_data: AsPathListRule = Field(..., description="Complete rule configuration")


class ReorderAsPathListRequest(BaseModel):
    """Model for reordering AS path list rules"""
    as_path_list_name: str = Field(..., description="AS path list name")
    rules: List[ReorderRuleItem] = Field(..., description="List of rules with new order")


class VyOSResponse(BaseModel):
    """Standard response from VyOS operations"""
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_as_path_list_capabilities(request: Request):
    """
    Get feature capabilities based on device VyOS version.

    Returns feature flags indicating which operations are supported.
    Allows frontends to conditionally enable/disable features.
    """
    return await load_vyos_capabilities(
        request=request,
        feature_group=FeatureGroup.BGP_AS_PATH,
        builder_cls=AsPathListBatchBuilder,
    )


# ============================================================================
# Endpoint 2: Config (Generalized Data)
# ============================================================================


@router.get("/config", response_model=AsPathListConfig)
async def get_as_path_list_config(http_request: Request, refresh: bool = False):
    """
    Get all AS path list configuration from VyOS in a generalized format.

    Args:
        refresh: If True, force refresh from VyOS. If False, use cache.

    Returns:
        Generalized configuration data optimized for frontend consumption
    """
    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        # Navigate to policy -> as-path-list
        as_path_list_config = full_config.get("policy", {}).get("as-path-list", {})

        if not as_path_list_config:
            return AsPathListConfig(as_path_lists=[], total=0)

        # Parse AS path lists
        as_path_lists = []
        for name, apl_data in as_path_list_config.items():
            as_path_list = parse_as_path_list(name, apl_data)
            as_path_lists.append(as_path_list)

        return AsPathListConfig(as_path_lists=as_path_lists, total=len(as_path_lists))

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


def parse_as_path_list(name: str, apl_data: dict) -> AsPathList:
    """Parse AS path list configuration from VyOS format."""
    description = apl_data.get("description")

    rules = []
    rules_raw = apl_data.get("rule", {})
    if rules_raw:
        for rule_num, rule_data in rules_raw.items():
            rule = parse_rule(int(rule_num), rule_data)
            rules.append(rule)

    return AsPathList(
        name=name,
        description=description,
        rules=sorted(rules, key=lambda r: r.rule_number)
    )


def parse_rule(rule_number: int, rule_data: dict) -> AsPathListRule:
    """Parse AS path list rule from VyOS format."""
    description = rule_data.get("description")
    action = rule_data.get("action", "permit")
    regex = rule_data.get("regex")

    return AsPathListRule(
        rule_number=rule_number,
        description=description,
        action=action,
        regex=regex
    )


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch")
async def as_path_list_batch_configure(http_request: Request, body: AsPathListBatchRequest):
    """
    Execute a batch of configuration operations.

    Allows multiple changes in a single VyOS commit for efficiency.
    """
    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = AsPathListBatchBuilder(version=version)

        # Process operations using inspect for dynamic method calls
        for operation in body.operations:
            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = list(sig.parameters.keys())

            # Build arguments dynamically
            args = []

            # Add AS path list name
            if "name" in params:
                args.append(body.name)

            # Add rule number if specified and method accepts it
            if body.rule_number and "rule" in params:
                args.append(str(body.rule_number))

            # Add operation value if provided
            if operation.value and len(params) > len(args):
                # Check remaining parameters
                remaining_params = params[len(args):]
                for param in remaining_params:
                    if param != "self":
                        args.append(operation.value)
                        break

            method(*args)

        # Execute batch
        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "Configuration updated"},
            error=response.error if response.error else None
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# Reorder Endpoint
# ============================================================================


@router.post("/reorder")
async def reorder_as_path_list_rules(http_request: Request, body: ReorderAsPathListRequest):
    """
    Reorder AS path list rules by deleting and recreating them in a single commit.

    This endpoint efficiently reorders multiple rules by:
    1. Deleting all specified rules in reverse order
    2. Recreating them with new rule numbers
    All operations are executed in a single VyOS commit.

    Args:
        request: Reorder request containing AS path list name and list of rules

    Returns:
        VyOSResponse with success/failure information
    """
    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = AsPathListBatchBuilder(version=version)

        # Step 1: Delete all rules in reverse order
        rules_to_delete = sorted([r.old_number for r in body.rules], reverse=True)
        for old_number in rules_to_delete:
            builder.delete_rule(body.as_path_list_name, str(old_number))

        # Step 2: Recreate rules with new numbers
        for rule_item in body.rules:
            new_number = rule_item.new_number
            rule_data = rule_item.rule_data

            # Create the rule
            builder.set_rule(body.as_path_list_name, str(new_number))

            # Set action
            if rule_data.action:
                builder.set_rule_action(body.as_path_list_name, str(new_number), rule_data.action)

            # Set description
            if rule_data.description:
                builder.set_rule_description(body.as_path_list_name, str(new_number), rule_data.description)

            # Set regex
            if rule_data.regex:
                builder.set_rule_regex(body.as_path_list_name, str(new_number), rule_data.regex)

        # Execute batch
        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": f"Successfully reordered {len(body.rules)} rules in AS path list {body.as_path_list_name}"},
            error=response.error if response.error else None
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
